src/rules.jl

```diff
@@ -93,6 +95,133 @@ julia> hypot_pullback(1) == (NO_FIELDS, (x / hypot(x, y)), (y / hypot(x, y)))
 true
 ```
 
+When defining overloads they should be wrapped with the [`@rrule`](@ref) macro.
+
 See also: [`frule`](@ref), [`@scalar_rule`](@ref)
 """
 rrule(::Any, ::Vararg{Any}; kwargs...) = nothing
+
+"""
+    @frule(function ...)
+
+[`frule`](@ref) defining functions should be decorated with this macro.
+
+Example:
+```julia
+@frule function frule((Δself, Δargs...), ::typeof(foo), args...; kwargs...)
+    ...
+    return y, ∂y
+end
+```
+"""
+macro frule(ast)
+    return quote
+        $(esc(ast))
+        $_register_new_rule(frule, $(QuoteNode(ast)))
+    end
+end
+
+"""
+    @rrule(function ...)
+
+[`rrule`](@ref) defining functions should be decorated with this macro.
+
+Example:
+```julia
+@rrule function rrule(::typeof(foo), args...; kwargs...)
+    ...
+    return y, pullback
+end
+```
+"""
+macro rrule(ast)
+    return quote
+        $(esc(ast))
+        $_register_new_rule(rrule, $(QuoteNode(ast)))
+    end
+end
+
+"""
+    _rule_list(frule | rrule)
+
+Returns a list of all rules currently defined rules of the given kind.
+"""
+_rule_list
+_rule_list(::typeof(rrule)) = RRULES
+_rule_list(::typeof(frule)) = FRULES
+const SIGT = Type
+const FRULES = Vector{Pair{SIGT, Expr}}()
+const RRULES = Vector{Pair{SIGT, Expr}}()
+
+
+function _register_new_rule(rule_kind, ast)
+    method = _just_defined_method(rule_kind)
+    rule_sig::SIGT = method.sig 
+    sig = _primal_sig(rule_kind, rule_sig)
+    push!(_rule_list(rule_kind), sig=>ast)
+    _trigger_new_rule_hooks(rule_kind, sig, ast)
+    return nothing
+end
+
+function _primal_sig(::typeof(frule), rule_sig)
+    @assert rule_sig.parameters[1] == typeof(frule)
+    # need to skip frule and the deriviative info, so starting from the 3rd
+    return Tuple{rule_sig.parameters[3:end]...}
+end
+
+function _primal_sig(::typeof(rrule), rule_sig)
+    @assert rule_sig.parameters[1] == typeof(rrule)
+    # need to skip rrule so starting from the 2rd
+    return Tuple{rule_sig.parameters[2:end]...}
```

Comment:
Does this work properly for kwargs? Or, rather, as properly as any of our stuff does for kwargs?

Reply:
> Does this work properly for kwargs? Or, rather, as properly as any of our stuff does for kwargs?

Yes it does.
Because kwargs don't show up in the signature.
Our stuff works great for kwargs.
Just accept a `kwargs...` and pass it on to the `frule` or `rrule`